File: 24_swap_nodes_in_pairs/main.py
# -*- coding: utf-8 -*-


class ListNode(object):

    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):

    def swap_pairs(self, head):
        pre, pre.next = self, head
        while pre.next and pre.next.next:
            a = pre.next
            b = a.next
            pre.next, b.next, a.next = b, a, b.next
            pre = a
        return self.next


def output(l):
    p = l
    while p:
        print(p.val),
        p = p.next
    print('')
    

if __name__ == '__main__':
    node1 = ListNode(1)
    node2 = ListNode(2)
    node1.next = node2
    node3 = ListNode(3)
    node2.next = node3
    node4 = ListNode(4)
    node3.next = node4

    s = Solution()
    output(node1)
    node1 = s.swap_pairs(node1)
    output(node1)

File: 24_swap_nodes_in_pairs/go/main.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func output(l *ListNode) {
	for p := l; p != nil; p = p.Next {
		fmt.Printf("%d ", p.Val)
	}
	fmt.Println()
}

func swapPairs(head *ListNode) *ListNode {
	pp := &head
	for *pp != nil && (*pp).Next != nil {
		a := *pp
		b := a.Next
		a.Next = b.Next
		b.Next = a
		*pp = b
		pp = &(a.Next)
	}
	return head
}

func main() {
	node1 := &ListNode{Val: 1}
	node2 := &ListNode{Val: 2}
	node1.Next = node2
	node3 := &ListNode{Val: 3}
	node2.Next = node3
	node4 := &ListNode{Val: 4}
	node3.Next = node4

	output(node1)
	node1 = swapPairs(node1)
	output(node1)
}
